### PR TITLE
Add test for pointwise broadcast TMA example

### PR DESCRIPTION
**Pull request by rdspring1**
Diagram of a pointwise kernel that use TMA loads for its input arguments and vectorized store for its output.

### CODE REVIEW

File: tests/cpp/test_tutorial.cpp [hunk 1]
```diff
@@ -697,4 +697,201 @@ TEST_F(Tutorial, IdModelReshapeAnalysis) {
...
+
+  // Vectorization
+  inlineSelectedAt({tv2b}, tv2, -2);
+
+  // Inline all operations except caching
+  inlineSelectedAt(
+      {all_tvs_except_cache.begin(), all_tvs_except_cache.end()},
+      reference_tv,
+      /*reference_pos=*/-1,
+      /*best_effort=*/true);
```
Comment (zasdfgbnm) [inline tests/cpp/test_tutorial.cpp:875]: If we add a `|| id->getParallelType() == ParallelType::Bulk` to
https://github.com/NVIDIA/Fuser/blob/0fff0d1ffe15524f1bc77dfc09f24b267c7ffea0/csrc/inlining.cpp#L69-L73
Can we then just `inlineMost()`?

File: tests/cpp/test_tutorial.cpp [hunk 2]
```diff
@@ -697,4 +697,201 @@ TEST_F(Tutorial, IdModelReshapeAnalysis) {
...
+TEST_F(Tutorial, PointwiseBroadcastTMA) {
+  CompileParams index32bit{DataType::Int32, 255, false};
+
+  auto fusion = std::make_unique<Fusion>();
+  FusionGuard fg(fusion.get());
+
+  constexpr at::ScalarType dtype = at::ScalarType::Float;
+
+  auto tv0 = makeContigTensor(3, aten_to_data_type(dtype));
+  auto tv1 = makeContigTensor(4, aten_to_data_type(dtype));
```
Comment (zasdfgbnm) [inline tests/cpp/test_tutorial.cpp:709]: To be consistent with https://github.com/NVIDIA/Fuser/pull/2130/files#diff-d21d811df5193bfc259b20940ca4e5fe43c6648a136c201f7884dbce4decbe2dR448-R449, can we create tensor with contiguity
```
T0[I1, I2, I3] contiguity: T, T, T
T1[I0, I1, I2, I3] contiguity: T, F, T, T
```

File: tests/cpp/test_tutorial.cpp [hunk 3]
```diff
@@ -697,4 +697,201 @@ TEST_F(Tutorial, IdModelReshapeAnalysis) {
...
+  //         split: [I0, I2, 256]
+  reference_tv->split(-1, tma_tile);
+  //         split: [I0, I3, 128, 2]
+  reference_tv->split(-1, vectorization);
+  //         split: [I4, 2, I3, 128, 2]
+  reference_tv->split(0, num_stages);
+  //         split: [I5, 132, 2, I3, 128, 2]
+  reference_tv->split(0, num_ctas_for_hopper);
+  //         reorder: [I5, I3, 132, 2, 128, 2]
+  reference_tv->reorder({{3, 1}, {1, 2}, {2, 3}});
```
Comment (zasdfgbnm) [inline tests/cpp/test_tutorial.cpp:835]: Looks like the names and order of IterDomain here mismatch with the plot you create? Also, should `num_stages` be `4` according to your plot?

File: tests/cpp/test_tutorial.cpp [hunk 4]
```diff
@@ -697,4 +697,201 @@ TEST_F(Tutorial, IdModelReshapeAnalysis) {
...
+  // Transform Operations between cache operations and output reference
+  TransformPropagator propagator(reference_tv);
+  MaxRootDomainInfoSpanningTree(reference_tv).traverse(&propagator);
+
+  reference_tv->axis(2)->parallelize(ParallelType::BIDx);
+  scheduler_utils::parallelizeAllLike(reference_tv);
+
+  // Vectorization for writing results to gmem
+  reference_tv->axis(-2)->parallelize(ParallelType::TIDx);
+  reference_tv->axis(-1)->parallelize(ParallelType::Vectorize);
```
Comment (zasdfgbnm) [inline tests/cpp/test_tutorial.cpp:846]: Should we also parallelize `-3` as `Unroll` and not inlining it?

File: tests/cpp/test_tutorial.cpp (hunk 3 above)
Comment (rdspring1) [reply tests/cpp/test_tutorial.cpp:835]: Yes, I made some changes in the plot for clarity but didn't update the test yet. 
- I thought having `num_stages` and `vectorization` both be 2 is confusing. 
- Reordering 3 iterDomains complicates the plot, so I changed it to swap [4 and I3].

File: tests/cpp/test_tutorial.cpp [hunk 5]
```diff
@@ -1250,4 +1250,190 @@ TEST_F(Tutorial, BasicTMA) {
...
+  //   root domain: [I0, I1]
+
+  constexpr int64_t num_threads = 128;
+  constexpr int64_t vectorization = 2;
+  constexpr int64_t tma_tile = num_threads * vectorization;
+  constexpr int64_t num_stages = 4;
+  constexpr int64_t num_ctas_for_hopper = 132;
+
+  // After TMA domain creation
+  //         split: [I0, I2, 256]
```
Comment (zasdfgbnm) [inline tests/cpp/test_tutorial.cpp:1384]: Should this be [I0, I3, 256]?

File: tests/cpp/test_tutorial.cpp [hunk 6]
```diff
@@ -1250,4 +1250,190 @@ TEST_F(Tutorial, BasicTMA) {
...
+  reference_tv->split(-1, tma_tile);
+  //         split: [I0, I3, 128, 2]
+  reference_tv->split(-1, vectorization);
+  //         split: [I4, 4, I3, 128, 2]
```
Comment (zasdfgbnm) [inline tests/cpp/test_tutorial.cpp:1388]: Should the first dim be I2?

File: tests/cpp/test_tutorial.cpp [hunk 7]
```diff
@@ -1250,4 +1250,190 @@ TEST_F(Tutorial, BasicTMA) {
...
+  reference_tv->split(0, num_stages);
+  //         split: [I5, 132, 4, I3, 128, 2]
```
Comment (zasdfgbnm) [inline tests/cpp/test_tutorial.cpp:1390]: Should the first dim be I4?

**Comment by zasdfgbnm** [comment]
For the picture, should the compute at position of T0a, T1a, and T2b be 3?

File: doc/dev/tma.md [hunk 8]
```diff
@@ -475,6 +477,28 @@ outputs:
 T1[I1, I0] contiguity: T, T
 ```
 
-Schedule:
+schedule:
+
+![example 3: bank-conflict-free transpose](tma/example3.svg)
+
+#### Example 4: tma-load inputs and vectorize-store output pointwise kernel
```
[the remark this answers is not in the document]
Comment (rdspring1) [reply doc/dev/tma.md:484]: Ah, I misinterpreted the comment.

File: doc/dev/tma.md [hunk 9]
```diff
@@ -475,6 +499,6 @@ outputs:
 T1[I1, I0] contiguity: T, T
 ```
 
-Schedule:
+schedule:
```
Comment (zasdfgbnm) [inline doc/dev/tma.md:502]: Could you revert the change in this area?

File: doc/dev/tma.md [hunk 10]
```diff
@@ -439,8 +439,32 @@ Please go to [test_tutorial.cpp](../../tests/cpp/test_tutorial.cpp) and search `
...
+math:
+T2[I0, I1] = T0[I0, I1] + T1[I0, I1];
+
+outputs:
+T2[I0, I1] contiguity: T, T
+```
+
+schedule:
+
+![example 1: tma-load-inputs, vectorize-store-output pointwise](tma/example1.svg)
```
Comment (zasdfgbnm) [inline doc/dev/tma.md:462]: nit:
`Schedule` instead of `schedule`, and `Example 1: TMA-...` instead of `example 1: tma-...`

File: tests/cpp/test_tutorial.cpp [hunk 11]
```diff
@@ -1250,4 +1250,204 @@ TEST_F(Tutorial, BasicTMA) {
...
+
+  // Propagate common parallel dimensions
+  reference_tv->axis(1)->parallelize(ParallelType::BIDx);
+  scheduler_utils::parallelizeAllLike(reference_tv);
+
+  // Vectorization for writing results to gmem
+  reference_tv->axis(-3)->parallelize(ParallelType::Unroll);
+  reference_tv->axis(-2)->parallelize(ParallelType::TIDx);
+  reference_tv->axis(-1)->parallelize(ParallelType::Vectorize);
+
```
Comment (zasdfgbnm) [inline tests/cpp/test_tutorial.cpp:1420]: Should we also do `tv2b->axis(-2)->parallelize(ParallelType::TIDx);`? The generated code does not look right to me:

```C++
...
      for(nvfuser_index_t i30 = 0; i30 < 4; ++i30) {
        nvfuser_index_t i31;
        i31 = 256LL * i30;
        #pragma unroll
        for(nvfuser_index_t i32 = 0; i32 < 128; ++i32) {
          nvfuser_index_t i33;
          i33 = i31 + (2LL * i32);
          #pragma unroll
          for(nvfuser_index_t i34 = 0; i34 < 2; ++i34) {
            nvfuser_index_t i35;
            i35 = i33 + i34;
            T5[i35]
              = T3[i35]
              + T4[i35];
          }
        }
      }
...
```

I think you should either inline to position -2, or manually parallelize it as TIDx

File: doc/dev/tma/example1.svg
Comment (zasdfgbnm) [inline doc/dev/tma/example1.svg:1]: Should the allocation of T2b also include the `4`?

File: tests/cpp/test_tutorial.cpp [hunk 12]
```diff
@@ -1250,4 +1250,204 @@ TEST_F(Tutorial, BasicTMA) {
   }
 }
 
+TEST_F(Tutorial, PointwiseBroadcastTMA) {
```
Comment (zasdfgbnm) [inline tests/cpp/test_tutorial.cpp:1253]: Could you add a `NVFUSER_TEST_CUDA_ARCH_GUARD(9, 0);` to both tests, and move this test below `VectorizeStorePointwiseTMA`?

File: doc/dev/tma/example1.svg [hunk 13]
```diff
@@ -0,0 +1 @@
...AADPUExURQAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAJljpgQAAABEdFJOUwAECAwQFBgcICQoLDA0ODxAREhMUFRYXGBkaGxwdHh8gIOHi4+Tl5ufo6Snq6+zt7u/w8fLz9PX297f4+fr7/P09/v+voEZfQAAAAlwSFlzAAAywAAAMsABKGRa2wAACi9JREFUeF7tXYdy2zgQleQSp/eeXHq79MSJc7Etn3X4/286kHiLBSSLBIgFwJnozdycACbiC8uWtwtossEGG2ywwQZ/JmZP3j7fxudxYOe3Ump+EaNR4IVmpNRHjEaBk5aSGtGtu24YqT2MRwBz35TawngE+GwY/cRwDDg2lF5gOALsGUbqIcYjwG1QuobxCPASlHYxHgG+GEZHGI4BPwylrxiOAUejo3TOMFJ/YzwC0Av37w1M1McTUFLz0fg4sgFKfcFMdbwFIY2xRCffwEepxVgoHYKQUocjiU62wOc//d9TzFXGDcOoxXwcd+456LR4i8m6+A42BtcxWxXwcMAPzNbEFFwItzBfERTlEkZgwa+Cyj/4/2IHB+rhHqhYT/cIB+rhGZhcbZSKBkczHKmGd2Cya73vXRypBpil+eSm+aDUBxypBmS6+5PJgfmk5lMcqoQd8NCX5iE+1k4xr4DG88nkAj6qNzhWCfdB44H+jHuojuveOZKWGndLL5+6bY5VwiewOKc/38Fn9d0cqwS8ZifN561TM1DqcnusEsDBxCT2ztUUvygOeN+O6PVTv9phHVwCB1yWXxiq82ZcAyQvPzFDev9oXAMUmtwzQ2JY8849BgWIJjOUCXSsYiYqgDKmSxi/wbhi9kRBEimn9s6dVHMqUE4VhpMp+bl6CR2M928MnTtXLRyAC2GnZmPLQ0yUBimnxng3mFk/V+mdoyTuGcYaqDq1QV0NXMPp72Os8QBTtawlUYLxbkD3slYIfhdn
```
Comment (zasdfgbnm) [inline doc/dev/tma/example1.svg:1]: Could you add a white background to this picture? Just add a `style="background-color:white"` to `<svg>`. Also, could you add copyright info? See https://github.com/NVIDIA/Fuser/blob/4d45883df7cb6656f8b099c794aa740e527f1348/doc/dev/tma/arbitrary-view.svg#L1-L6

File: tests/cpp/test_tutorial.cpp [hunk 14]
```diff
@@ -1250,4 +1250,210 @@ TEST_F(Tutorial, BasicTMA) {
...
+  FusionExecutor fe;
+  fe.compileFusion(fusion.get(), {at_tv0, at_tv1}, {}, index32bit);
+  auto outputs = fe.runFusion({at_tv0, at_tv1});
+
+  auto at_output = at_tv0 + at_tv1;
+  testValidate(
+      fusion.get(), outputs, {at_tv0, at_tv1}, {at_output}, __LINE__, __FILE__);
+}
+
+TEST_F(Tutorial, VectorizeStorePointwiseTMA) {
```
Comment (zasdfgbnm) [inline tests/cpp/test_tutorial.cpp:1357]: Move this above `PointwiseBroadcastTMA`?

File: doc/dev/tma/example1.svg
Comment (zasdfgbnm) [inline doc/dev/tma/example1.svg:1]: In the picture, could you mark the 128 of T2b as TIDx?

**Review by zasdfgbnm — APPROVED** [review]
Thanks for creating these examples!
